File: 2020/04/day04.cc
#include <cstdio>
#include <unordered_map>
#include <vector>

using namespace std;

const vector<long long> splits(long long n) {
  vector<long long> ret;
  ret.push_back(n);
  long long rhs = 0;
  long long lhs = n;
  long long ten_pow = 1;
  while (true) {
    rhs = (lhs % 10) * ten_pow + rhs;
    ten_pow *= 10;
    lhs /= 10;

    if (lhs == 0) break;

    auto sums = splits(lhs);
    for (int i = 0; i < sums.size(); ++i) {
      ret.push_back(sums[i] + rhs);
    }
  }

  return ret;
}

int main() {
  const long long N = 1000000000000;
  // const int N = 10000;

  long long sum = 0;
  for (long long i = 4; i * i <= N; ++i) {
    auto sums = splits(i * i);
    for (int j = 0; j < sums.size(); ++j) {
      if (sums[j] == i) {
        sum += i * i;
        break;
      }
    }
  }

  printf("%lld\n", sum);

  return 0;
}

File: 2020/04/day04.rb
require 'set'

N = 1_000_000_000_000
# N = 10_000

def splits(n)
  @memo ||= {}
  return @memo[n] if @memo.key?(n)

  @memo[n] = [n].to_set
  rhs = 0
  lhs = n
  ten_pow = 1
  loop do
    rhs = (lhs % 10) * ten_pow + rhs
    ten_pow *= 10
    lhs /= 10

    break if lhs == 0

    @memo[n].merge(splits(lhs).map { |s| s + rhs })
  end

  @memo[n]
end

answer = []
(4..Math.sqrt(N)).each do |i|
  answer << i * i if splits(i * i).include?(i)
end

puts answer.sum
